src/DIRAC/Resources/IdProvider/OAuth2IdProvider.py

```diff
@@ -224,70 +366,57 @@ def getScopeGroups(self, scope: str) -> list[str]:
-                self.jwks_fetch_last = time.time()
-                return S_OK(self.jwks)
-            except Exception as e:
-                self.log.exception(e)
-                return S_ERROR(f"Error {repr(e)}")
-        return S_OK()
+    #############################################################################
 
-    def verifyToken(self, accessToken=None, jwks=None):
+    def verifyToken(self, accessToken):
         """Verify access token
 
         :param str accessToken: access token
         :param dict jwks: JWKs
 
         :return: dict
         """
-        # Define an access token
         if not accessToken:
-            accessToken = self.token["access_token"]
-        # Renew a JWKs of an identity provider if needed
-        if not jwks:
-            result = self.updateJWKs()
-            if not result["OK"]:
-                return result
-            jwks = self.jwks
-        if not jwks:
-            return S_ERROR("JWKs not found.")
+            return S_ERROR("Access token is empty")
+
+        # Get JWKs
+        result = self.getJWKs()
+        if not result:
+            return S_ERROR("JWKs not found")
+        jwks = result
+
         # Try to decode and verify an access token
         self.log.debug(f"Try to decode token {accessToken} with JWKs:\n", pprint.pformat(jwks))
         try:
             return S_OK(jwt.decode(accessToken, JsonWebKey.import_key_set(jwks)))
-        except Exception as e:
-            self.log.exception(e)
-            return S_ERROR(repr(e))
+        except DecodeError as e:
+            return S_ERROR(f"The provided token cannot be decoded: {e}")
+
+    #############################################################################
 
     @gRefreshToken
-    def refreshToken(self, refresh_token=None, group=None, **kwargs):
+    def refreshToken(self, refreshToken, group=None, **kwargs):
         """Refresh token
 
         :param str token: refresh_token
         :param str group: DIRAC group
 
         :return: dict
         """
+        if not refreshToken:
+            return S_ERROR("Refresh token is empty")
         if group:
             # If group set add group scopes to request
             if not (groupScopes := self.getGroupScopes(group)):
                 return S_ERROR(f"No scope found for {group}")
 
             kwargs.update(dict(scope=list_to_scope(groupScopes)))
 
-        if not refresh_token:
-            refresh_token = self.token.get("refresh_token")
         try:
-            token = self.refresh_token(self.get_metadata("token_endpoint"), refresh_token=refresh_token, **kwargs)
+            token = self.refresh_token(self.get_metadata("token_endpoint"), refresh_token=refreshToken, **kwargs)
             return S_OK(OAuth2Token(dict(token)))
-        except Exception as e:
-            self.log.exception(e)
-            return S_ERROR(repr(e))
+        except OAuthError as e:
```

Comment:
Are you sure there can not be other exceptions than this specific one ? They will pass uncaught if there is no general exception catching